server/main.go

```diff
@@ -129,50 +118,82 @@ func main() {
 				errChan <- fmt.Errorf("http_plain server error: %v", err)
 			}
 		}()
 	}
 
-	err := <-errChan // this waits for some server breaking
-	log.Fatalf("Error: %v", err)
+	err = <-errChan // this waits for some server breaking
+	log.WithError(err).Fatalf("Fail")
+}
+
+// httpsBouncerHandler decides what kind of requests it is and redirects to GRPC if needed.
+func httpsBouncerServer(grpcHandler *grpc.Server, httpHandler http.Handler, logEntry *log.Entry) *http.Server {
+	httpBouncerHandler := http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		if req.URL.Path == "/_healthz" {
+			healthEndpoint(w, req)
+			return
+		}
+		if strings.HasPrefix(req.Header.Get("content-type"), "application/grpc") {
+			grpcHandler.ServeHTTP(w, req)
+			return
+		}
+		httpHandler.ServeHTTP(w, req)
+	}).ServeHTTP
+
+	return &http.Server{
+		WriteTimeout: *flagHttpMaxWriteTimeout,
+		ReadTimeout:  *flagHttpMaxReadTimeout,
+		ErrorLog:     http_logrus.AsHttpLogger(logEntry.WithField(ctxtags.TagsForScheme, "tls")),
+		Handler:      http.HandlerFunc(httpBouncerHandler),
+	}
+}
+
+func debugServer(logEntry *log.Entry) *http.Server {
+	return &http.Server{
+		WriteTimeout: *flagHttpMaxWriteTimeout,
+		ReadTimeout:  *flagHttpMaxReadTimeout,
+		ErrorLog:     http_logrus.AsHttpLogger(logEntry.WithField(ctxtags.TagsForScheme, "tls")),
+		Handler: chi.Chain(
+			http_ctxtags.Middleware("debug"),
+			http_debug.Middleware(),
+			http_logrus.Middleware(logEntry.WithField(ctxtags.TagsForScheme, "plain")),
+		).Handler(debugMux()),
+	}
 }
 
-func registerDebugHandlers() {
+func debugMux() *chi.Mux {
+	m := chi.NewMux()
 	// TODO(mwitkow): Add middleware for making these only visible to private IPs.
-	http.Handle("/_healthz", http.HandlerFunc(healthEndpoint))
-	http.Handle("/debug/metrics", prometheus.UninstrumentedHandler())
-	http.Handle("/debug/flagz", http.HandlerFunc(flagz.NewStatusEndpoint(sharedflags.Set).ListFlags))
-	//http.Handle("/debug/pprof/", http.HandlerFunc(pprof.Index))
-	//http.Handle("/debug/pprof/cmdline", http.HandlerFunc(pprof.Cmdline))
-	//http.Handle("/debug/pprof/profile", http.HandlerFunc(pprof.Profile))
-	//http.Handle("/debug/pprof/symbol", http.HandlerFunc(pprof.Symbol))
-	//http.Handle("/debug/pprof/trace", http.HandlerFunc(pprof.Trace))
-	//http.Handle("/debug/events", http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-	//	w.Header().Set("Content-Type", "text/html; charset=utf-8")
-	//	trace.Render(w, req /*sensitive*/, true)
-	//}))
-	//http.Handle("/debug/events", http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-	//	w.Header().Set("Content-Type", "text/html; charset=utf-8")
-	//	trace.RenderEvents(w, req /*sensitive*/, true)
-	//}))
+	m.Handle("/_healthz", http.HandlerFunc(healthEndpoint))
+	m.Handle("/debug/metrics", prometheus.UninstrumentedHandler())
+	m.Handle("/debug/flagz", http.HandlerFunc(flagz.NewStatusEndpoint(sharedflags.Set).ListFlags))
```

Comment:
We can either make private IP filtering or just include OIDC authorizer to this, what do you think?